### PR TITLE
fix: setupControllerConnection outstream end event listener

### LINKED ISSUE #26002: Outstream on 'end' callback, in `setupControllerConnection` of MM controller, is not called on UI close
This problem was discovered when investigating https://github.com/MetaMask/metamask-extension/issues/25951

In `metamask-controller.js` there is a method `setupControllerConnection`. It sets up a listener for the end of the outstream like this:
```
    outStream.on('end', () => {
      this.activeControllerConnections -= 1;
      this.emit(
        'controllerConnectionChanged',
        this.activeControllerConnections,
      );
      this.removeListener('update', handleUpdate);
    });
```

As of https://github.com/MetaMask/metamask-extension/pull/24533, this callback is not called when the UI is closed.

To repro:
1. Checkout the commit of that PR 72d4b5265c
2. Add a debugger or console.log inside the aforementioned callback
3. `yarn start` and then install and onboard
4. Close the UI. The debugger or console log will not be hit

But if those steps are repeated on the prior commit, 928d30afcd, the debugger or console.log will be hit

### PR DESCRIPTION
**Pull request by danjm**
## **Description**

This PR addresses a problem which had the following surface level symptoms: when the UI is closed, the extension continues to send network requests to the provider.

Upon some investigation, we discovered that these requests were the result of polling initiated from the account tracker. And then upon further investigation, we discovered that the callback in the `outstream` `'end'` listener in `setupControllerConnection` of `metamask-controller.js` was not being called when the UI was closed.

This problem was introduced with https://github.com/MetaMask/metamask-extension/pull/24533. That PR updates the `object-multiplex` dependency to v2.0.0. The major change in that update is the update of the `readable-stream` dependency from a v2 version to a v3 version. With that update, we see new `premature-close` errors in the background. It seems that these errors are still present in v2, but they are handled differently, resulting in an `'error'` event instead of an `'end'` event. We still have not established a definitive understanding of why this is happening, and follow up investigation will be needed.

However, this PR still puts forward a fix that safely resolves the issue. The solution added here is to make use of the `readableStream.finished` function. The callback passed to this function will be called when the stream is ended. This function has the same behaviour as the `outstream.on('end'` listener had prior to the aforementioned PR.

Instead of fully replacing the existing listener, we have added the `readableStream.finished` call in addition to the existing listener. The reason for this is explained in a code comment:

> The presence of both of the below handlers may be redundant. After upgrading metamask/object-multiples to v2.0.0, which included an upgrade of readable-streams from v2 to v3, we saw that the `outStream.on('end'` handler was almost never being called. This seems to related to how v3 handles errors vs how v2 handles errors; there are "premature close" errors in both cases, although in the case of v2 they don't prevent `outStream.on('end'` from being called. At the time that this comment was committed, it was known that we need to investigate and resolve the underlying error, however, for expediency, we are not addressing them at this time. Instead, we can observe that `readableStream.finished` preserves the same functionality as we had when we relied on readable-stream v2. Meanwhile, the `outStream.on('end'` handler was observed to have been called at least once. As we do not yet fully understand the conditions under which both of these are being called, we include both handlers at the risk of redundancy. Logic has been added to the handler to ensure that calling it more than once does not have any affect.

To achieve "Logic has been added to the handler to ensure that calling it more than once does not have any affect.", the PR adds a unique id to each stream passed to `setupControllerConnection`, and uses a `Set` to ensure the handler code only runs once per stream.

This PR then adds automated tests. The added unit tests are focused on validating both pre-existing and new behaviour of `setupControllerChanged`. The added e2e tests focus on the inititially mentioned surface level symptoms: it confirms that `eth_call` and `eth_getBalance` calls, the calls coming from the account-tracker initiated polling, are not continued after the UI is closed. This e2e test fails on current develop but passes on this branch.

MetaMask team members can see more details in this thread: https://consensys.slack.com/archives/CTQAGKY5V/p1721653870583869

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/26141?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/26002

## **Manual testing steps**

To observe the bug on develop:
1. Install and onboard
2. Open the service worker dev console and go to the network tab
3. Observe `eth_call` and `eth_getBalance` requests to infura
4. Close the UI, observe that the requests from step 3 continue

On this branch, in step 4, after closing the UI, those requests will not continue.

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: privacy-snapshot.json
Comment (Gudahtt) [inline privacy-snapshot.json:1]: For a later PR: We should consider sorting this with prettier (`prettier-plugin-sort-json` can do it)

File: privacy-snapshot.json [hunk 1]
```diff
@@ -23,18 +24,22 @@
   "gas.api.cx.metamask.io",
   "github.com",
   "goerli.infura.io",
+  "lattice.gridplus.io",
+  "localhost:12345",
```
Comment (Gudahtt) [inline privacy-snapshot.json:28]: Note for other reviewers, these two localhost entries appear to be the only additions. The other changes are sorting and removing duplicates.

File: test/e2e/webdriver/chrome.js [hunk 2]
```diff
@@ -30,6 +30,7 @@ class ChromeDriver {
   }) {
     const args = [
       `--proxy-server=${getProxyServer(proxyPort)}`, // Set proxy in the way that doesn't interfere with Selenium Manager
+      `--proxy-bypass-list=<-loopback>`,
```
[the remark this answers is not in the document]
Comment (danjm) [reply test/e2e/webdriver/chrome.js]: Oh, actually, this is dead code. Now that these tests run on mainnet and I am mocking the infura calls, we don't need to rely on checking network requests to localhost. So I removed this and the associated privacy snapshot update in aac4e571ab

File: privacy-snapshot.json [hunk 3]
```diff
@@ -23,18 +24,22 @@
...
   "localhost:8000",
   "localhost:8545",
```
[the remark this answers is not in the document]
Comment (danjm) [reply privacy-snapshot.json]: Removed in latest commit. See https://github.com/MetaMask/metamask-extension/pull/26141#discussion_r1692272446

**Review by legobeat — DISMISSED** [review]

File: test/e2e/tests/api-usage/account-tracker-api-usage.spec.ts [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        driver: Driver;
+        mockedEndpoint: MockedEndpoint[];
+      }) => {
+        await unlockWallet(driver);
+        await driver.delay(3000);
+        const initialInfuraJsonRpcRequests = await getAllInfuraJsonRpcRequests(
+          mockedEndpoint,
+        );
+
+        await driver.openNewURL('chrome://extensions');
```
Comment (seaona) [inline test/e2e/tests/api-usage/account-tracker-api-usage.spec.ts:148]: It looks like this makes the firefox test fail. Could we selectively go to 

```
if (process.env.SELENIUM_BROWSER === 'chrome') {
  await driver.openNewURL('chrome://extensions');
}
if (process.env.SELENIUM_BROWSER === 'firefox') {
  await driver.openNewURL('about:debugging#/runtime/this-firefox');
}   
```

File: test/e2e/tests/api-usage/account-tracker-api-usage.spec.ts [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+          jsonrpc: '2.0',
+          id: '1111111111111111',
+          result: '0x0',
+        },
+      };
+    }),
+  ];
+}
+
+async function getAllInfuraJsonRpcRequests(
```
[the remark this answers is not in the document]
Comment (seaona) [reply test/e2e/tests/api-usage/account-tracker-api-usage.spec.ts:56]: +1!

File: test/e2e/tests/api-usage/account-tracker-api-usage.spec.ts [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        allInfuraJsonRpcRequests.push(json);
+      }
+    }
+  }
+
+  return allInfuraJsonRpcRequests;
+}
+
+describe('Account Tracker API Usage', function () {
+  it('should not make eth_call or eth_getBalance requests before the UI is opened and should make those requests after the UI is opened', async function () {
```
Comment (seaona) [inline test/e2e/tests/api-usage/account-tracker-api-usage.spec.ts:81]: I see that the account tracker is also responsible from initiating `eth_getBlockByNumber` calls. Should we also include them here?

https://github.com/user-attachments/assets/871e7182-0221-40c3-a80c-10f68ad8b6b9

File: test/e2e/tests/api-usage/account-tracker-api-usage.spec.ts (hunk 4 above)
Comment (danjm) [reply test/e2e/tests/api-usage/account-tracker-api-usage.spec.ts]: I fixed it by replacing the url with `about:blank` https://github.com/MetaMask/metamask-extension/pull/26141/commits/a31544443deed3f068d5ec04857d22a6ca82fc0d

Comment (seaona) [reply test/e2e/tests/api-usage/account-tracker-api-usage.spec.ts]: oh nice :heart_eyes: ty!!

File: test/e2e/tests/api-usage/account-tracker-api-usage.spec.ts (hunk 6 above)
Comment (danjm) [reply test/e2e/tests/api-usage/account-tracker-api-usage.spec.ts:94]: good suggestion, added here ac50f666ee

**Comment by chloeYue** [comment]
QA test OK !

File: test/e2e/tests/api-usage/account-tracker-api-usage.spec.ts (hunk 6 above)
Comment (seaona) [reply test/e2e/tests/api-usage/account-tracker-api-usage.spec.ts:94]: thank you Dan, great refactor! Could it be that we are missing it [here](https://github.com/MetaMask/metamask-extension/commit/ac50f666eebae7403308e438d4d4598d0a762778#diff-174d8183603b0004a7ff4acb5a21374ea2ad88e63ab433090783ef41db9c35e7R86) and in the test description (not a blocker though)?

Comment (danjm) [reply test/e2e/tests/api-usage/account-tracker-api-usage.spec.ts:94]: No, in that case there is an eth_getBlockbyNumber call before the UI opens, for other reasons. That should be addressed separately.

I will create a ticket for that and add a code comment and update the test description

File: test/e2e/tests/api-usage/account-tracker-api-usage.spec.ts [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+      },
+      async ({
+        driver,
+        mockedEndpoint,
+      }: {
+        driver: Driver;
+        mockedEndpoint: MockedEndpoint[];
+      }) => {
+        await unlockWallet(driver);
+        await driver.delay(3000);
```
Comment (DDDDDanica) [inline test/e2e/tests/api-usage/account-tracker-api-usage.spec.ts:170]: Nit: can we use predefined delay here 
```suggestion
        await driver.delay(veryLargeDelayMs);
```

File: test/e2e/tests/api-usage/account-tracker-api-usage.spec.ts [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+        const initialInfuraJsonRpcRequests = await getAllInfuraJsonRpcRequests(
+          mockedEndpoint,
+        );
+
+        await driver.openNewURL('about:blank');
+        await driver.delay(20000);
```
Comment (DDDDDanica) [inline test/e2e/tests/api-usage/account-tracker-api-usage.spec.ts:176]: same as https://github.com/MetaMask/metamask-extension/pull/26141/files#r1695290161

Comment (danjm) [reply test/e2e/tests/api-usage/account-tracker-api-usage.spec.ts:183]: we need this to be 20000 ms long. I have added a code comment to explain 27df50c86d

**Review by seaona — APPROVED** [review]
tests look good, though might be good to have a dev approval too, for the controllers change

**Review by DDDDDanica — APPROVED** [review]

**Pull request merged**